timemachine/cpp/src/nonbonded_all_pairs.cu

```diff
@@ -173,6 +175,25 @@ void NonbondedAllPairs<RealType>::set_atom_idxs_device(
     // Force the rebuild of the nblist
     gpuErrchk(cudaMemsetAsync(d_rebuild_nblist_, 1, 1 * sizeof(*d_rebuild_nblist_), stream));
     this->K_ = K;
+    // Reset the steps so that we do a new sort
+    this->steps_ = 0;
+}
+
+template <typename RealType> bool NonbondedAllPairs<RealType>::needs_sort() { return steps_ % STEPS_PER_SORT == 0; }
+
+template <typename RealType>
+void NonbondedAllPairs<RealType>::sort(const double *d_coords, const double *d_box, cudaStream_t stream) {
+    // (ytz): update the permutation index before building neighborlist, as the neighborlist is tied
+    // to a particular sort order which means having to rebuild the neighborlist
+    if (!disable_hilbert_) {
+        this->hilbert_sort(d_coords, d_box, stream);
+    } else {
+        gpuErrchk(cudaMemcpyAsync(
+            d_sorted_atom_idxs_, d_atom_idxs_, K_ * sizeof(*d_atom_idxs_), cudaMemcpyDeviceToDevice, stream));
+    }
+    gpuErrchk(cudaMemsetAsync(d_rebuild_nblist_, 1, sizeof(*d_rebuild_nblist_), stream));
+    // Set the pinned memory to indicate that we need to rebuild
+    p_rebuild_nblist_[0] = 1;
```

Comment:
Do we need to trigger a nblist rebuild if not Hilbert sorting?

Reply:
In the edge case of changing the atom indices we do (which resets steps to 0), though generally we don't.

I avoided optimizing this path as it is only used for debugging to my knowledge.

Reply:
More generally, if not Hilbert sorting, could we just do

```cpp
        gpuErrchk(cudaMemcpyAsync(
            d_sorted_atom_idxs_, d_atom_idxs_, K_ * sizeof(*d_atom_idxs_), cudaMemcpyDeviceToDevice, stream));
```

once during initialization?

Reply:
> In the edge case of changing the atom indices we do

In this case, could the reinitialization be done as part of `set_atom_idxs`?

Reply:
It could be, but that would seem to add complexity to optimize a path that we rarely use and only for debugging.

Reply:
OK. Agreed that it's probably not worth optimizing for the case `disable_hilbert=True`. I'd only advocate these changes if they make the code easier to understand. (In the specific case of `set_atom_idxs`, it seems a bit safer/clearer to explicitly reinitialize `d_sorted_atom_idxs` rather than rely on this happening later because `steps==0`, but defer to your judgement)

Reply:
Because the hilbert sort is dependent on coords and box for its sort, you wouldn't be able to build a good sort from `set_atom_idxs`. You could use the last stored coords in `d_nblist_coords_`, but that seems inconsistent.